File: android/app/src/main/java/com/easybc/planner/sync/shared/ProfileLabels.kt
package com.easybc.planner.sync.shared

import com.keyneom.synckit.sharing.SharingRole

private const val MAX_DATASET_ID_LENGTH = 40

fun slugifyDatasetId(displayName: String): String {
    val slug = displayName
        .trim()
        .lowercase()
        .replace(Regex("[^a-z0-9]+"), "-")
        .trim('-')
        .take(MAX_DATASET_ID_LENGTH)
    require(slug.isNotEmpty() && slug != PRIMARY_DATASET_ID) {
        "Choose a profile name with letters or numbers."
    }
    return slug
}

fun uniqueOwnedDatasetId(
    displayName: String,
    ownerEmail: String,
    profiles: List<ProfileRecord>,
): String {
    val base = slugifyDatasetId(displayName)
    val ownedIds = profiles
        .filter {
            it.ownerEmail.equals(ownerEmail, ignoreCase = true) &&
                it.role.equals(SharingRole.OWNER.name, ignoreCase = true)
        }
        .map { it.datasetId }
        .toSet()
    if (!ownedIds.contains(base)) return base
    var suffix = 2
    while (ownedIds.contains("$base-$suffix")) suffix += 1
    val candidate = "$base-$suffix"
    require(candidate.length <= MAX_DATASET_ID_LENGTH) {
        "Too many profiles with similar names. Choose a different name."
    }
    return candidate
}

/** Opaque storage identity for new profiles; the mutable label is metadata. */
fun newOwnedDatasetId(
    existingDatasetIds: Iterable<String>,
    randomUuid: () -> String = { java.util.UUID.randomUUID().toString() },
): String {
    val existing = existingDatasetIds.toSet()
    repeat(10) {
        val candidate = "p-${randomUuid().lowercase()}"
        if (candidate.length <= MAX_DATASET_ID_LENGTH && candidate !in existing) return candidate
    }
    error("Could not create a unique profile identifier. Try again.")
}

fun isOwnedProfile(state: SharedSyncState, profile: ProfileRecord): Boolean =
    profile.ownerEmail.equals(state.ownerEmail, ignoreCase = true) &&
        profile.role.equals(SharingRole.OWNER.name, ignoreCase = true)

fun profileDisplayLabel(state: SharedSyncState, profile: ProfileRecord): String {
    if (isLocalProfile(profile)) {
        return profile.displayName?.trim()?.takeIf { it.isNotEmpty() } ?: "Local profile"
    }
    profile.localDisplayName?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
    profile.displayName?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
    if (!isOwnedProfile(state, profile)) return profile.folderName
    if (profile.datasetId == PRIMARY_DATASET_ID) return "My data"
    return profile.datasetId
}

fun disambiguatedProfileLabel(state: SharedSyncState, profile: ProfileRecord): String {
    val label = profileDisplayLabel(state, profile)
    val duplicates = state.profiles.count {
        profileDisplayLabel(state, it).equals(label, ignoreCase = true)
    }
    return if (duplicates > 1) "$label — ${profile.ownerEmail}" else label
}

fun findOwnedPrimaryProfile(state: SharedSyncState): ProfileRecord? =
    state.profiles.firstOrNull {
        isOwnedProfile(state, it) && it.datasetId == PRIMARY_DATASET_ID
    }

fun findOwnedStorageProfile(state: SharedSyncState): ProfileRecord? =
    findOwnedPrimaryProfile(state) ?: state.profiles.firstOrNull {
        isOwnedProfile(state, it) && !it.appFolderId.isNullOrBlank()
    }

File: web/src/sync/profileLabels.ts
import {
  isLocalProfile,
  PRIMARY_DATASET_ID,
  type ProfileRecord,
  type SharedSyncState,
} from "./sharedTypes";

const MAX_DATASET_ID_LENGTH = 40;

export function slugifyDatasetId(displayName: string): string {
  const slug = displayName
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, MAX_DATASET_ID_LENGTH);
  if (!slug || slug === PRIMARY_DATASET_ID) {
    throw new Error("Choose a profile name with letters or numbers.");
  }
  return slug;
}

export function uniqueOwnedDatasetId(
  displayName: string,
  ownerEmail: string,
  profiles: ProfileRecord[],
): string {
  const base = slugifyDatasetId(displayName);
  const ownedIds = new Set(
    profiles
      .filter(
        (profile) =>
          profile.ownerEmail.toLowerCase() === ownerEmail.toLowerCase() &&
          profile.role === "owner",
      )
      .map((profile) => profile.datasetId),
  );
  if (!ownedIds.has(base)) return base;
  let suffix = 2;
  while (ownedIds.has(`${base}-${suffix}`)) suffix += 1;
  const candidate = `${base}-${suffix}`;
  if (candidate.length > MAX_DATASET_ID_LENGTH) {
    throw new Error("Too many profiles with similar names. Choose a different name.");
  }
  return candidate;
}

/** New profile identity is opaque; labels are mutable metadata, not storage keys. */
export function newOwnedDatasetId(
  existingDatasetIds: Iterable<string>,
  randomUUID: () => string = () => crypto.randomUUID(),
): string {
  const existing = new Set(existingDatasetIds);
  for (let attempt = 0; attempt < 10; attempt += 1) {
    const candidate = `p-${randomUUID().toLowerCase()}`;
    if (candidate.length <= MAX_DATASET_ID_LENGTH && !existing.has(candidate)) {
      return candidate;
    }
  }
  throw new Error("Could not create a unique profile identifier. Try again.");
}

export function isOwnedProfile(state: SharedSyncState, profile: ProfileRecord): boolean {
  return (
    profile.ownerEmail.toLowerCase() === state.ownerEmail.toLowerCase() &&
    profile.role === "owner"
  );
}

export function profileDisplayLabel(state: SharedSyncState, profile: ProfileRecord): string {
  if (isLocalProfile(profile)) {
    return profile.displayName?.trim() || "Local profile";
  }
  if (profile.localDisplayName?.trim()) return profile.localDisplayName.trim();
  if (profile.displayName?.trim()) return profile.displayName.trim();
  if (isOwnedProfile(state, profile)) {
    if (profile.datasetId === PRIMARY_DATASET_ID) return "My data";
    return profile.datasetId;
  }
  return profile.folderName;
}

export function disambiguatedProfileLabel(
  state: SharedSyncState,
  profile: ProfileRecord,
): string {
  const label = profileDisplayLabel(state, profile);
  const duplicates = state.profiles.filter(
    (candidate) =>
      profileDisplayLabel(state, candidate).toLocaleLowerCase() === label.toLocaleLowerCase(),
  );
  return duplicates.length > 1 ? `${label} — ${profile.ownerEmail}` : label;
}

export function findOwnedPrimaryProfile(
  state: SharedSyncState,
): ProfileRecord | undefined {
  return state.profiles.find(
    (profile) =>
      isOwnedProfile(state, profile) && profile.datasetId === PRIMARY_DATASET_ID,
  );
}

export function findOwnedStorageProfile(
  state: SharedSyncState,
): ProfileRecord | undefined {
  return findOwnedPrimaryProfile(state) ??
    state.profiles.find((profile) => isOwnedProfile(state, profile) && profile.appFolderId);
}
